### PR TITLE
[CLI] add support for Recipes (squashed commits)

### PR DESCRIPTION
**Pull request by damianham**
### Description of the Change
Adds support to generate applications and scaffolding from recipes.  Recipes are stored in the amberframework/recipes repository.

This PR replaces #705, #709, #710 and #716 with a squashed commit log.

### Alternate Designs
None considered
### Benefits
Users will be able to generate apps with varying configs and different, shards and shard versions,
Users will be able to use different tooling easily e.g npm or yarn, webpack or browserify, gulp or grunt etc.

### Possible Drawbacks

There will be a multitude of app flavours so it might be more difficult to support users as they will have a wide variety of app configuration and features.

### CODE REVIEW

**Comment by eliasjpr** [comment]
@faustinoaq maybe would be good to add the `kind:feature` label this way we can start following some of the suggestions here https://github.com/amberframework/amber/issues/724

Also consider adding a `kind:Breaking Change` to keep track of breaking change PRs

**Comment by damianham** [comment]
@robocarp sorry about that.  Following the advice given in issue 724 I thought it was cleaner to tidy up the old PRs that were superceded by this one.  You can of course still access the old PRs #705, #709, #710 and #716 to review the code and I can reopen them if you wish.  I agree with @eliasjpr that it is better if the commit log has one sensible descriptive commit message for the feature rather than lots of  1 liners like 'fixed styling for ameba etc.' 

Most of the changes have already been approved in #710.  The difference between #710 and #716 is essentially this commit https://github.com/amberframework/amber/pull/716/commits/c477ad3714d5b698de924e2b6d4758252bbd29f2

**Comment by damianham** [comment]
@elorest I also submitted a PR for the online docs and amberframework.org to add documentation about this recipe feature to the online documentation.  The README.md file in amberframework/recipes also has some information about how to use it and how to create a recipe.

**Comment by damianham** [comment]
@faustinoaq is it a breaking change ?  I don't think so - the existing generator works as is if you don't specify a recipe name.

**Comment by damianham** [comment]
@robocarp RE: your previous question about filters in #710 
"Given that this code is written as a shim until the liquid project gets this filter, will this code break when the liquid project is updated?"
I have submitted the additional filters as a PR to the Techmagister/liquid.cr project.  I will keep on top of that and instigate the removal of the filters from this project once the PR is merged.

**Comment by elorest** [comment]
@damianham Thanks. Could respond as to the necessity of liquid as well?

**Comment by robacarp** [comment]
@damianham no worries about the new pull request, I appreciate your motive. In the future, it's usually easy to squash commits when merging a pull request and that allows discussion to flow easy enough. I think it's better to get the whole feature in one pull anyway.

image

File: shard.yml [hunk 1]
```diff
@@ -30,6 +30,9 @@ dependencies:
     github: jeromegn/kilt
     version: ~> 0.4.0
 
+  liquid:
+      github: TechMagister/liquid.cr
```
[the remark this answers is not in the document]
Comment (faustinoaq) [reply shard.yml]: Also indentation should be fixed here :sweat_smile:  (use 2 spaces)

[the remark this answers is not in the document]
Comment (damianham) [reply shard.yml]: Yes I have been trying to get TechMagister to tag liquid.cr with the tag v0.2.4 for days, my last email to him to add the tag v0.2.4 raterh than 0.2.4 was 2 days ago.  That is why I suggested we maintain a fork in amberframework so we can manage the releases better.

[the remark this answers is not in the document]
Comment (faustinoaq) [reply shard.yml]: Done! https://github.com/TechMagister/liquid.cr/issues/29 :sparkles: :clap: :+1:

[the remark this answers is not in the document]
Comment (faustinoaq) [reply shard.yml]: I guess an opening issue was the best option to contact him :wink:

**Pull request merged**